File: paths-in-matrix-whose-sum-is-divisible-by-k.py
class Solution(object):
    MOD = 10**9 + 7

    def numberOfPaths(self, grid, k):
        m, n = len(grid), len(grid[0])

        # dp[i][j][sum_mod]
        dp = [[[-1] * k for _ in range(n)] for _ in range(m)]

        def dfs(i, j, sum_mod):
            sum_mod = (sum_mod + grid[i][j]) % k

            # Base case
            if i == m - 1 and j == n - 1:
                return 1 if sum_mod == 0 else 0

            if dp[i][j][sum_mod] != -1:
                return dp[i][j][sum_mod]

            down = right = 0
            if i + 1 < m:
                down = dfs(i + 1, j, sum_mod)
            if j + 1 < n:
                right = dfs(i, j + 1, sum_mod)

            dp[i][j][sum_mod] = (down + right) % self.MOD
            return dp[i][j][sum_mod]

        return dfs(0, 0, 0)

File: paths-in-matrix-whose-sum-is-divisible-by-k.java
class Solution {

    private static int MOD = 1_000_000_007;

    public int numberOfPaths(int[][] grid, int k) {
        int n = grid.length, m = grid[0].length;
        int[][][] dp = new int[n][m][k];

        // Initialise dp[0][0][k].
        dp[0][0][grid[0][0] % k] = 1;

        // Initialise first column dp[i][0][k].
        int vj = grid[0][0] % k;
        for (int i = 1; i < n; i++) {
            vj += grid[i][0];
            vj %= k;
            dp[i][0][vj] = 1;
        }

        // Initialise first row dp[0][j][k].
        int vi = grid[0][0] % k;
        for (int j = 1; j < m; j++) {
            vi += grid[0][j];
            vi %= k;
            dp[0][j][vi] = 1;
        }

        // Solve dp.
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                for (int x = 0, p = (x + grid[i][j]) % k; x < k; x++, p++) {
                    if (p >= k) {
                        p -= k;
                    }
                    dp[i][j][p] = (dp[i - 1][j][x] + dp[i][j - 1][x]) % MOD;
                }
            }
        }
        return dp[n - 1][m - 1][0];
    }
}
